koku/koku/trino_database.py

```diff
@@ -78,6 +84,130 @@ def query_id(self) -> t.Optional[str]:
 
 
+class NoSuchKeyViolation:
+    """Detect NoSuchKey violation in an exception or error message."""
+
+    NO_SUCH_KEY_REGEX_STR = r"NoSuchKey"
+    NO_SUCH_KEY_REGEX = re.compile(NO_SUCH_KEY_REGEX_STR)
+
+    def __init__(self, _exception):
+        """Accepts Exception or str"""
+        self.message = _exception if isinstance(_exception, str) else str(_exception)
+        self.__is_no_such_key_violation = bool(self.NO_SUCH_KEY_REGEX.search(self.message))
+
+    @property
+    def is_no_such_key_violation(self):
+        """Returns bool indicating if NoSuchKey violation was detected."""
+        return self.__is_no_such_key_violation
+
+    def __bool__(self):
+        return self.__is_no_such_key_violation
+
+    def __repr__(self):
+        return str(self)
+
+    def __str__(self):
+        if self.__is_no_such_key_violation:
+            return f"NoSuchKey error detected: {self.message}"
+        else:
+            return "No NoSuchKey error detected."
+
+
+class TrinoNoSuchKeyException(Exception):
+    """Custom exception for NoSuchKey errors"""
+
+    pass
+
+
+def extract_context_from_sql_params(sql_params: dict[str, t.Any]) -> dict[str, t.Any]:
+    ctx = {}
+    if sql_params is None:
+        return ctx
+    if schema := sql_params.get("schema"):
+        ctx["schema"] = schema
+    if (start := sql_params.get("start")) or (start := sql_params.get("start_date")):
+        ctx["start_date"] = start
+    if (end := sql_params.get("end")) or (end := sql_params.get("end_date")):
+        ctx["end_date"] = end
+    if invoice_month := sql_params.get("invoice_month"):
+        ctx["invoice_month"] = invoice_month
+    if provider_uuid := sql_params.get("source_uuid"):
+        ctx["provider_uuid"] = provider_uuid
+    if cluster_id := sql_params.get("cluster_id"):
+        ctx["cluster_id"] = cluster_id
+
+    return ctx
+
+
+def retry(
+    original_callable: t.Callable = None,
+    *,
+    retries: int = settings.HIVE_PARTITION_DELETE_RETRIES,
+    retry_on: type[Exception] | tuple[type[Exception], ...] = TrinoExternalError,
+    max_wait: int = 30,
+    log_message: t.Optional[str] = "Retrying...",
+):
+    """Decorator with the retry logic."""
+
+    def _retry(callable: t.Callable):
+        @functools.wraps(callable)
+        def wrapper(*args, **kwargs):
+            context = kwargs.get("context", extract_context_from_sql_params(kwargs.get("sql_params", {}))) or None
+            for attempt in range(retries + 1):
+                try:
+                    LOG.debug(f"Attempt {attempt + 1} for {callable.__name__}")
+                    return callable(*args, **kwargs)
+
+                except retry_on as ex:
+                    LOG.debug(f"Exception caught: {ex}")
+
+                    # Verify if the exception contains "NoSuchKey"
+                    # If it does, retry the operation
+                    violation = NoSuchKeyViolation(ex)
+                    if violation.is_no_such_key_violation and attempt < retries:
```

Comment:
This evaluation shouldn't be done here. If this is called with `retry_on=TrinoNoSuchKeyException`, other exception types won't be retried, which is the end result we're aiming for.